src/java/com/unifina/datasource/DataSourceEventQueue.java

```diff
@@ -50,80 +63,110 @@ public void addDayListener(IDayListener dayListener) {
 	}
 
 	/**
-	 * The call to this method should block until the queue is aborted or all events have been processed.
+	 * The call to this method blocks until the queue is aborted or all events have been processed.
 	 */
 	public void start() throws Exception {
 		abort = false;
-		doStart();
+		asyncExecutor = new ThreadPoolExecutor(1, 1,
+			0L, TimeUnit.MILLISECONDS,
+			new LinkedBlockingQueue<>());
+
+		consumingThread = Thread.currentThread();
+
+		try {
+			runEventLoopUntilDone();
+		} finally {
+			asyncExecutor.shutdownNow();
+
+			// Report statistics
+			log.debug(eventQueueMetrics.toString());
+		}
 	}
 
-	public void enqueue(FeedEvent event) {
-		if (syncLock != null) {
-			synchronized (syncLock) {
-				event.queueTicket = queueTicket++;
-				if (queue.size() <= QUEUE_HARD_LIMIT) {
-					queue.add(event);
-				} else {
-					log.warn("Queue hard limit reached: " + event.toString());
-				}
-				syncLock.notify(); // Notify the SignalPathRunner thread
-			}
+	/**
+	 * Enqueues the event into this event queue.
+	 *
+	 * If the method is called from the same Thread which is consuming
+	 * messages from the queue, the event will be async-queued as a failsafe
+	 * to avoid deadlocking the thread in case the queue is full.
+	 */
+	public void enqueue(Event event) {
+		if (Thread.currentThread() == consumingThread) {
+			enqueueAsync(event);
 		} else {
-			event.queueTicket = queueTicket++;
-			queue.add(event);
+			enqueueSync(event);
 		}
 	}
 
-	public void abort() {
-		abort = true;
-		if (syncLock != null) {
-			synchronized (syncLock) {
-				syncLock.notify(); // Notify the SignalPathRunner thread
+	/**
+	 * Enqueues an event synchronously. Blocks if the queue is full, until the
+	 * event fits into the queue. Throws if no space becomes available before a timeout.
+	 */
+	private void enqueueSync(Event event) {
+		try {
+			boolean success = queue.offer(event, 30, TimeUnit.SECONDS);
+			if (!success) {
+				throw new RuntimeException("Timed out while trying to enqueue event " + event);
 			}
-			doStop();
+		} catch (InterruptedException e) {
+			log.error(e);
 		}
 	}
 
-	protected abstract Queue<FeedEvent> initQueue();
-
-	protected abstract void doStart() throws Exception;
+	/**
+	 * Non-blocking method to asynchronously queue events.
+	 * This is useful for internal event queuing, i.e. cases where
+	 * we want to eventually put something in the event queue even if
+	 * it's congested with incoming data.
+	 */
+	private void enqueueAsync(Event event) {
+		asyncExecutor.submit(() -> {
```

Comment:
It is only used in the special case that the producing thread (the thread in which `DataSourceEventQueue#enqueue()` is called) is the same as the consuming thread. This happens when handling an event causes new events to be queued.

It should never block the thread, because if it does, no events are consumed from the queue, and the whole process is deadlocked.

I changed it to a bounded queue, and added error logging in case it is full. When full, the events-to-be-queued will be dropped, which is the same behavior as with the main (sync) queue if the `offer` times out.